projects/RabbitMQ.Client/client/impl/Channel/AutorecoveringChannel.cs

```diff
@@ -0,0 +1,364 @@
+            => NonDisposedDelegate.CommitTransactionAsync();
+
+        public ValueTask RollbackTransactionAsync()
+            => NonDisposedDelegate.RollbackTransactionAsync();
+
+        public ValueTask ResendUnackedMessages(bool requeue)
+            => NonDisposedDelegate.ResendUnackedMessages(requeue);
+
+        public ValueTask CloseAsync(ushort replyCode, string replyText)
+        {
+            _connection.UnregisterChannel(this);
+            return NonDisposedDelegate.CloseAsync(replyCode, replyText);
+        }
+
+        public ValueTask AbortAsync(ushort replyCode, string replyText)
+        {
+            _connection.UnregisterChannel(this);
+            return NonDisposedDelegate.AbortAsync(replyCode, replyText);
+        }
+
+        public bool WaitForConfirms(TimeSpan timeout) => NonDisposedDelegate.WaitForConfirms(timeout);
+
+        public bool WaitForConfirms() => NonDisposedDelegate.WaitForConfirms();
+
+        public void WaitForConfirmsOrDie() => NonDisposedDelegate.WaitForConfirmsOrDie();
+
+        public void WaitForConfirmsOrDie(TimeSpan timeout) => NonDisposedDelegate.WaitForConfirmsOrDie(timeout);
+
+        private async ValueTask RecoverStateAsync()
+        {
+            if (_prefetchCountConsumer != 0)
+            {
+                await SetQosAsync(0,  _prefetchCountConsumer, false).ConfigureAwait(false);
+            }
+
+            if (_prefetchCountGlobal != 0)
+            {
+                await SetQosAsync(0,  _prefetchCountGlobal, true).ConfigureAwait(false);
+            }
+
+            if (_usesPublisherConfirms)
+            {
+                await ActivatePublishTagsAsync().ConfigureAwait(false);
+            }
+
+            if (_usesTransactions)
+            {
+                await ActivateTransactionsAsync().ConfigureAwait(false);
+            }
+        }
+
+        private void RunRecoveryEventHandlers()
+        {
+            ThrowIfDisposed();
+            foreach (EventHandler<EventArgs> reh in Recovery?.GetInvocationList() ?? Array.Empty<Delegate>())
+            {
+                try
+                {
+                    reh(this, EventArgs.Empty);
+                }
+                catch (Exception e)
+                {
+                    _delegate.OnUnhandledExceptionOccurred(e, "OnModelRecovery");
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ThrowIfDisposed()
+        {
+            if (_disposed)
+            {
+                ThrowDisposed();
+            }
+        }
+
+        public override string? ToString() => NonDisposedDelegate.ToString();
+
+        public async ValueTask DisposeAsync()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            await NonDisposedDelegate.DisposeAsync().ConfigureAwait(false);
+            _disposed = true;
+        }
+
+        void IDisposable.Dispose()
```

Comment:
I don't see much value in implementing both `IDisposable` as well as `IAsyncDisposable`. Especially not because the first one requires us to do evil sync over async. Let's embrace async properly in v8

Reply:
We can think about dropping IDisposable, but there might be some drawbacks with it (e.g. DI container might not fully support it or 🤷‍♂️ )
I do not know what's the best way forward for this.